static/js/util/profile_edit.js

```diff
@@ -214,6 +215,53 @@ export function boundCheckbox(keySet: string[], label: string|React$Element<*>):
   );
 }
 
+
+export function boundGeosuggest(keySet: string[], label: string|React$Element<*>,
+  { id, placeholder, types }: { id: string, placeholder: string, types: string[] } = {}
+): React$Element<*> {
+  const {
+    profile,
+    errors,
+    updateProfile,
+    validator,
+    updateValidationVisibility,
+  } = this.props;
+
+  const addressComponentTypeMap = {
+    // From the name that we use, to the component type that Google Maps uses.
+    // See https://developers.google.com/maps/documentation/geocoding/intro#Types
+    "city": "locality",
+    "state_or_territory": "administrative_area_level_1",
+    "country": "country",
+  };
+
+  const onSuggestSelect = (suggest) => {
+    let clone = _.cloneDeep(profile);
+    const components = suggest.gmaps.address_components;
+    keySet.forEach((key) => {
+      const gmapType = addressComponentTypeMap[key];
+      const component = _.find(components, (c) => _.includes(c.types, gmapType));
+      clone[key] = component.long_name;
+    });
+    updateValidationVisibility(keySet);
+    updateProfile(clone, validator);
+  };
+
+  const initial = _.join(_.map(keySet, (key) => profile[key]), ', ');
+
+  return (
+    <div>
+      <Geosuggest initialValue={initial}
+        id={id} label={label} placeholder={placeholder} types={types}
+        onSuggestSelect={onSuggestSelect}
+      />
+      <span className="validation-error-text">
+        {_.get(errors, keySet)}
+      </span>
+    </div>
+  );
+}
```

Comment:
A couple things in this function could probably be pulled out as little helper methods - if it's feasible to do that I think it's a good idea (especially for things like `addressComponentTypeMap`) because otherwise a bunch of things are being unnecessarily redeclared on every function call.